src/Url.php

```diff
@@ -2,70 +2,298 @@
-            array_merge(
-                self::URL_DEFAULT_PARTS,
-                (array)parse_url($src)
-            )
-        );
-
-        if ($scheme === 'data') {
+        // Bail early on 'data:' assets.
+        if ($this->scheme === 'data') {
             return false;
         }
 
-        $userpass = $user;
+        // @TODO: This probably needs additional logic.
+
+        $parts = array_merge(self::URL_DEFAULT_PARTS, $this->toArray(true));
+
+        return (bool)filter_var(self::toString($parts), FILTER_VALIDATE_URL);
+    }
+
+    /**
+     * Get the URL parts as an associative array.
+     *
+     * @param bool $sparse Whether to only include parts with non-empty values.
+     *
+     * @return array Associative array with URL parts.
+     */
+    public function toArray($sparse = false)
+    {
+        $array = [
+            self::SCHEME   => $this->scheme,
+            self::HOST     => $this->host,
+            self::PORT     => $this->port,
+            self::USER     => $this->user,
+            self::PASS     => $this->pass,
+            self::PATH     => $this->path,
+            self::QUERY    => $this->query,
+            self::FRAGMENT => $this->fragment,
+        ];
+
+        return $sparse ? array_filter($array) : $array;
+    }
+
+    /**
+     * Serialize to string.
+     *
+     * @return string URL.
+     */
+    public function __toString()
+    {
+        return self::toString($this->toArray());
+    }
+
+    /**
+     * Return a provided URL parsed into parts as an assembled string.
+     *
+     * @param array $parts Parts of the URL to assemble.
+     * @return string Assembled URL string.
+     */
+    public static function toString($parts)
+    {
+        $url = ! empty($parts[self::SCHEME]) ? "{$parts[self::SCHEME]}://" : '//';
+
+        if (! empty($parts[self::USER])) {
+            $url .= "{$parts[self::USER]}";
+            if (! empty($parts[self::PASS])) {
+                $url .= ":{$parts[self::PASS]}";
+            }
+            $url .= '@';
+        }
+
+        $url .= ! empty($parts[self::HOST]) ? "{$parts[self::HOST]}" : 'localhost';
+
+        if (! empty($parts[self::PORT])) {
+            $url .= ":{$parts[self::PORT]}";
+        }
+
+        $url .= ! empty($parts[self::PATH]) ? "{$parts[self::PATH]}" : '/';
+
+        if (! empty($parts[self::QUERY])) {
+            $url .= "?{$parts[self::QUERY]}";
+        }
 
-        if (! empty($pass)) {
-            $userpass .= ":{$pass}";
+        if (! empty($parts[self::FRAGMENT])) {
+            $url .= "#{$parts[self::FRAGMENT]}";
         }
 
-        if (! empty($userpass)) {
-            $userpass .= '@';
+        return $url;
+    }
+
+    /**
+     * Magic getter to return the individual parts.
+     *
+     * @param string $name Name of the part to return.
+     * @return string|null Part string or null if it was not found during parsing.
+     */
+    public function __get($name)
```

Comment:
Eventually we should implement `__set()` as well, since it would be very useful to be able to change out just the `host`, when rewriting a URL to use the AMP Cache for example. 

Either this, or make the member variables `public`.

Reply:
I'd recommend against that. This is supposed to behave like a value object, so it should be immutable to avoid bugs.

It could have a method to retrieve a new clone with a different host, though:
```
$cachedUrl = $url->withHost($cachedHost);
```
The important part is that the original object is not changed (as that would changed it in all the places it was passed around), but rather we return a new instance with the adapted host.